sdks/python/apache_beam/io/gcp/dicomio.py

```diff
@@ -0,0 +1,419 @@
+    {
+      'success': boolean value telling whether the store is successful
+      'input': undeliverable data. Exactly the same as the input,
+        only set if the operation is failed.
+      'status': status code from the server, used as error messages.
+    }
+  """
+  def __init__(self, destination_dict, input_type, credential=None):
+    """Initializes DicomStoreInstance.
+    Args:
+      destination_dict: # type: python dict, encodes DICOM endpoint information:
+        {
+          'project_id': str,
+          'region': str,
+          'dataset_id': str,
+          'dicom_store_id': str,
+        }
+        Key-value pairs:
+          project_id: Id of the project in which DICOM store locates. (Required)
+          region: Region where the DICOM store resides. (Required)
+          dataset_id: Id of the dataset where DICOM store belongs to. (Required)
+          dicom_store_id: Id of the dicom store. (Required)
+      input_type: # type: string, could only be 'bytes' or 'fileio'
+      credential: # type: Google credential object, if it is specified, the
+        Http client will use it instead of the default one.
+    """
+    self.credential = credential
+    self.destination_dict = destination_dict
+    # input_type pre-check
+    if input_type not in ['bytes', 'fileio']:
+      raise ValueError("input_type could only be 'bytes' or 'fileio'")
+    self.input_type = input_type
+
+  def expand(self, pcoll):
+    return pcoll | beam.ParDo(
+        _StoreInstance(self.destination_dict, self.input_type, self.credential))
+
+
+class _StoreInstance(beam.DoFn):
+  """A DoFn read or fetch dicom files then push it to a dicom store."""
+  def __init__(self, destination_dict, input_type, credential=None):
+    self.credential = credential
+    # pre-check destination dict
+    required_keys = ['project_id', 'region', 'dataset_id', 'dicom_store_id']
+    for key in required_keys:
+      if key not in destination_dict:
+        raise ValueError('Must have %s in the dict.' % (key))
+    self.destination_dict = destination_dict
+    self.input_type = input_type
+
+  def process(self, element):
+    project_id = self.destination_dict['project_id']
+    region = self.destination_dict['region']
+    dataset_id = self.destination_dict['dataset_id']
+    dicom_store_id = self.destination_dict['dicom_store_id']
+
+    # Read the file based on different input. If the read fails ,return
+    # an error dict which records input and error messages.
+    dicom_file = None
+    try:
+      if self.input_type == 'fileio':
+        f = element.open()
+        dicom_file = f.read()
+      else:
+        dicom_file = element
+    except Exception as error_message:
+      error_out = {}
+      error_out['status'] = error_message
+      error_out['input'] = element
+      error_out['success'] = False
+      return [error_out]
+
+    # Feed the dicom file into store client
+    _, status_code = DicomApiHttpClient().dicomweb_store_instance(
```

Comment:
In this case it's also worth buffering and performing the requests in parallel.

Reply:
When you talked about buffering and applying parallelism did you assume that the file will be written to GCS? 
As I know there is no way of doing that in DICOM store but I will take a look if such optimizations exist.

Reply:
TODO(@George-Wu ) add feature

Reply:
@pabloem Feature added